src/api/java/com/minecolonies/api/util/InventoryUtils.java

```diff
@@ -1473,7 +1486,7 @@ public static boolean transferItemStackIntoNextFreeSlotInItemHandler(
                                                                            final int sourceIndex,
                                                                            @NotNull final IItemHandler targetHandler)
     {
-        ItemStack sourceStack = sourceHandler.extractItem(sourceIndex, Integer.MAX_VALUE, true);
+        ItemStack sourceStack = sourceHandler.extractItem(sourceIndex, Integer.MAX_VALUE, false);
```

Comment:
You are violating the IITemHandler contract here!

Always simulate first before performing any operation

Reply:
where do I find that IITemHandler contract? in the documentation I only see that input stacks are not to be modified, thats all

Reply:
It is not really a contract.
It is the general rule that was setup when they where defined:

First test if you actually can modify the inventory
Then actually the modification when you are happy with it.

Reply:
ah I see, so for that we always want to do extract simulate -> insert simulate -> extract real -> insert real?

Reply:
btw I saw that forge(or maybe vanilla?) does insert simulate -> set stack for inventories

Reply:
Both work really.
Important is that you do a simulate first.
Set stack just overrides.

Forge does it in vanilla cause insert simulate -> set stack basically is the same.

Reply:
Personally I would create something like an Inventory interaction handler
Which records somehow the interactions thgat you want to do.
THen returns the simulated results, and allows you to tell it to commit the change.

Reply:
ye that would be nice to have, kinda like DB transactions

Reply:
Yes exactly.
Something that has the exact same signatures as IItemHandler (Except maybe the simulate flag) or something.

Reply:
Shouldn't we able to code this in a utility function somehow?

Reply:
ye should be doable, although I'd leave that for the rework, since this was just supposed to be bugfixes